sync-core/src/main/java/com/cloudant/http/HttpConnection.java

```diff
@@ -0,0 +1,234 @@
+                                   URL url) {
+        this.requestMethod = requestMethod;
+        this.contentType = contentType;
+        this.url = url;
+        this.requestProperties = new HashMap<String, String>();
+    }
+
+    /**
+     * Set the String of data to be sent to the server
+     * @param input String of data to be sent to the server
+     */
+    public void setInput(final String input) {
+        this.input = new ByteArrayInputStream(input.getBytes());
+        // input is in bytes, not characters
+        this.inputLength = input.getBytes().length;
+    }
+
+    /**
+     * Set the byte array of data to be sent to the server
+     * @param input byte array of data to be sent to the server
+     */
+    public void setInput(final byte[] input) {
+        this.input = new ByteArrayInputStream(input);
+        this.inputLength = input.length;
+    }
+
+    /**
+     * Set the InputStream of data to be sent to the server
+     * @param input InputStream of data to be sent to the server
+     */
+    public void setInput(InputStream input) {
+        this.input = input;
+    }
+
+    /**
+     * Set the InputStream of data, of known length, to be sent to the server
+     * @param input InputStream of data to be sent to the server
+     * @param inputLength Length of data to be sent to the server, in bytes
+     */
+    public void setInput(InputStream input, long inputLength) {
+        this.input = input;
+        this.inputLength = inputLength;
+    }
+
+    private void executeInternal() throws IOException {
+        connection = (HttpURLConnection) url.openConnection();
+        for (String key : requestProperties.keySet()) {
+            connection.setRequestProperty(key, requestProperties.get(key));
+        }
+        // always read the result, so we can retrieve the HTTP response code
+        connection.setDoInput(true);
+        connection.setRequestMethod(requestMethod);
+        if (contentType != null) {
+            connection.setRequestProperty("Content-type", contentType);
+        }
+        if (username != null) {
+            String encodedAuth = Base64.encodeBase64String(
+                    String.format("%s:%s", username, password != null ? password : "")
+                            .getBytes());
+            connection.setRequestProperty("Authorization", String.format("Basic %s", encodedAuth));
+        }
+        if (input != null) {
+            connection.setDoOutput(true);
+            if (inputLength != -1) {
+                // TODO on 1.7 upwards this method takes a long, otherwise int
+                connection.setFixedLengthStreamingMode((int)this.inputLength);
+            } else {
+                // TODO some situations where we can't do chunking, like multipart/related
+                /// https://issues.apache.org/jira/browse/COUCHDB-1403
+                connection.setChunkedStreamingMode(1024);
+            }
+            connection.setRequestProperty("Expect", "100-continue");
```

Comment:
Be good to see a brief explanation/link to RFC for what the expected behaviour is here, and at least a "this header helps prevent us sending data when server would reject it (e.g., for bad credentials)"